EIPS/eip-5450.md

```diff
@@ -58,48 +58,50 @@ In the second validation phase control-flow analysis is performed on the code.
 
-The FIFO queue *worklist* is used for tracking "to be inspected" instructions. Initially, it contains the first instruction.
+The first instruction's recorded stack height bounds are initialized to be equal to the number of inputs to the function type matching the code (`stack_height_min = stack_height_max = type[code_section_index].inputs`).
 
-While *worklist* is not empty, dequeue an instruction and:
+For each instruction:
 
-1. Determine the effect the instruction has on the operand stack:
-   1. **Check** if the recorded stack height satisfies the instruction requirements. Specifically:
-      - for `CALLF` instruction the recorded stack height must be at least the number of inputs of the called function according to its type defined in the type section.
-      - for `RETF` instruction the recorded stack height must be exactly the number of outputs of the function matching the code.
-   2. Compute new stack height after the instruction execution.
-2. Determine the list of successor instructions that can follow the current instructions:
+1. **Check** if this instruction has recorded stack height bounds. If it does not, it means it was neither referenced by previous forward jump, nor is part of sequential instruction flow, and this code fails validation. 
+2. Determine the effect the instruction has on the operand stack:
+   1. **Check** if the recorded stack height bounds satisfy the instruction requirements. Specifically:
+      - for `CALLF` instruction the recorded stack height lower bound must be at least the number of inputs of the called function according to its type defined in the type section.
+      - for `RETF` instruction both the recorded lower and upper bound must be equal and must be exactly the number of outputs of the function matching the code.
+      - for `JUMPF` into returning function both the recorded lower and upper bound must equal exactly `type[current_section_index].outputs + type[target_section_index].inputs - type[target_section_index].outputs`.
+      - for `JUMPF` into non-returning function the recorded stack height lower bound must be at least the number of inputs of the target function according to its type defined in the type section.
+      - for any other instruction the recodrded stack height lower bound must be at least the number of inputs required by instruction
+      -  there is no additional check for terminating instructions other than `RETF` and `JUMPF`, this implies that extra items left on stack at instruction ending EVM execution are allowed.
+   2. For `CALLF` and `JUMPF` **check** for possible stack overflow: if recorded stack height upper bound is greater than `1024 - types[target_section_index].max_stack_height + types[target_section_index].inputs`, validation fails.
+   3. Compute new stack height bounds after the instruction execution. Upper and lower bound are updated by the same value.
```

Comment:
specify the value of the update, esp. for CALLF

Reply:
Added.